vault/expiration.go

```diff
@@ -279,63 +290,139 @@ func (m *ExpirationManager) Restore(errorFunc func(), loadDelay time.Duration) (
 	}
 	m.logger.Debug("expiration: leases collected", "num_existing", len(existing))
 
-	// Restore each key by pulling from the result chan
-	for i, leaseID := range existing {
-		if i > 0 && i%500 == 0 {
-			m.logger.Trace("expiration: leases loading", "progress", i)
+	// Make the channels used for the worker pool
+	broker := make(chan string)
+	quit := make(chan bool)
+	// Buffer these channels to prevent deadlocks
+	errs := make(chan error, len(existing))
+	result := make(chan *leaseEntry, len(existing))
+
+	// Use a wait group
+	wg := &sync.WaitGroup{}
+
+	// Create 64 workers to distribute work to
+	for i := 0; i < consts.ExpirationRestoreWorkerCount; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				select {
+				case leaseID, ok := <-broker:
+					// broker has been closed, we are done
+					if !ok {
+						return
+					}
+
+					// If we loaded a lease elsewhere in the expiration manager during
```

Comment:
Don't we need to be holding the lock lease before this restoreLoaded call? Otherwise we could check whether it was loaded, the by the time we lock it could be loaded.

Also, why send a nil down result...why not just continue?